protobuf-format.md

```diff
@@ -0,0 +1,387 @@
+
+```java
+CloudEvent.newBuilder()
+  .setCloudEventsVersion("1.0")
+  .setExtensions(
+    Struct.newBuilder()
+    .putFields(
+      "comExampleMyextension",
+      Value.newBuilder().setStringValue("myvalue").build()))
+  .build();
+```
+
+In the next CloudEvents minor release, the extension is promoted to an
+official field. The protobuf message definition is updated for
+CloudEvents 1.1:
+
+```java
+CloudEvent.newBuilder()
+  .setCloudEventsVersion("1.1")
+  .setComExampleMyextension("myvalue")
+  // During the upgrade process producers SHOULD continue setting
+  // exts bag to maintain backwards compat
+  .setExtensions(
+    Struct.newBuilder()
+      .putFields(
+        "comExampleMyextension",
+        Value.newBuilder().setStringValue("myvalue").build()))
+  .build();
+```
+
+#### 2.3.1 CloudEvents upgrade process
+
+If a field moves from the extensions bag to a top level field, then
+the producers and consumers of a CloudEvents system SHOULD coordinate
+in the upgrade process:
+
+1. Initially, the producers and consumers are using CloudEvents 1.0
+   and the extension is expressed in the “extensions” bag.
+1. CloudEvents 1.1 is released.
+1. The producers write the extension to both the extensions bag and
+   the type safe top level field. The messages will have
+   "cloudEventsVersion" set to "1.1", but the extension is still
+   readable by 1.0 consumers.
+1. All consumers upgrade to 1.1 and stop reading from the extensions
+   bag and switch to reading from only the type safe top level field.
+1. The producers stop writing to the extensions bag, and only write to
+   the type safe top level field.
+
+
+### 2.4 Relation to CloudEvents JSON format:
+
+All proto3 messages have a standard JSON form. The standard JSON of
+this protobuf format is not compatible from the official [CloudEvents
+JSON encoding][CE_JSON_ENCODING] at the time of writing. A service
+that supports the standard protobuf JSON encoding as well as the
+official CloudEvents JSON encoding will end up supporting two
+different JSON encodings.
+
+Below are a few examples of the proto3 JSON.
+
+A CloudEvent whose payload is a JSON may be constructed as follows:
+```java
+CloudEvent.newBuilder()
+  .setEventType("com.example.usercreated")
+  .setEventTypeVersion("1.0")
+  .setCloudEventsVersion("0.1")
+  .setSource("producer1")
+  .setEventId("100")
+  .setTimestamp(ts)
+  .setContentType("application/json")
+  .setSchemaUrl("https://com.example.schema/usercreated")
+  .setData(
+    Value.newBuilder().setStructValue(Struct.newBuilder()
+    .putFields(
+        "username",
+        Value.newBuilder().setStringValue("theusername").build())
+    .putFields(
+        "email",
+        Value.newBuilder().setStringValue("user@example.com").build())
+    .build()))
+  .setExtensions(
+    Struct.newBuilder()
+      .putFields(
+        "comExampleExtension1",
+        Value.newBuilder().setStringValue("value1").build())
+      .putFields(
+        "comExampleExtension2",
```

Comment:
Maybe it would add some clarity if the field name was `com_example_extension_2`?

If I get https://developers.google.com/protocol-buffers/docs/proto3#json correctly, the snake_case to camelCase conversion is applied to proto field names, but - I guess - the names within a `Struct` aren't touched, as they are not proto fields?

Reply:
I agree that, per convention, extension attributes should be in camelCase. However, this is not a MUST, and a snake_cased attribute must not be transformed to camelCase.

I guess everything works as expected after I dug into the proto json spec, but it wasn't immediately clear to me why e.g. `event_id` is turned into `eventId`, but `com_example_extension` isn't.

Reply:
I see what you mean, I think I misunderstood your original comment. In the case of extensions being in the `Struct`, fields can be arbitrarily cased because it is a map keyed on case sensitive strings.
If we ever introduce a top level attribute that is not lowerCamelCase then we must use the `json_name` option to override the json field name.

Reply:
Thanks for the clarification!